### DISCUSSION TITLE
Switching between remote and local proxy at runtime

### DISCUSSION THREAD

**Discussion by kobi-shanti** [MarimerLLC/csla#3533] [Questions]
Hi,

We just wanted to know if it is possible to switch between remote and local proxy at runtime? Using CSLA5 specifically.

Thanks in advance for any help.

**Comment by hurcane** [comment]
We might need more information regarding the context in which you ask this question. The data portal can be defined via a configuration file, which is read when the runtime process starts. The RunLocal allows specific methods to always run locally, regardless of the default data portal configuration.

In what scenario would you want to be able to change the location of the data portal call?

**Reply by kobi-shanti** [reply]
Thanks for the response Don. A simplified example of the context would be within a WinForms application, after startup of the application. We were hoping to be able to switch between local and remote proxy (specifically local and HTTP proxy) without changing configuration and restarting the application. 

The reason for wanting to do this is twofold. First we'd like to compare performance of our application in 2-tier vs 3-tier and we'd also potentially want to allow some of our users to run in 2-tier while allowing others to run in 3-tier.

**Reply by hurcane** [reply]
You need code somewhere that is going to determine whether calls should be made using the remote or local data portal. The key question is when is that decision made. Off the top of my head, I can think of three techniques where you could apply the conditional behavior. I've listed my ideas in what I think is order of increasing effort. These are all hypothetical suggestions.

### Startup
I think the easiest way to perform this experiment is to have custom code at application launch to determine which kind of portal you want to use for that process. Simply call a different configuration method. This could be ideal for controlling the data portal depending on the user identity.

### RunLocal
Another idea that comes to mind is to configure the remote data portal by default and using the RunLocal attribute on the data portal methods that you want to run without using http. If you want to test the same business object dynamically that way, you could use operator overloading and have two versions of the criteria class. For example, you would have the MyCriteria class that includes the necessary properties. Also have a MyCriterialLocal class that inherits from MyCrtieria with no additional behavior. The business object would have one method with the [Create] attribute that takes MyCriteria as the parameter. Another method that takes the MyCriteriaLocal paramter would have the [Create] and [RunLocal] attributes.

### Custom data portal classes
A third option might be to create some custom data portal channel classes that will encapsulate the logic that determines the type of data portal. I think you could apply the conditional logic within a class that implements IDataPortalProxy.IsServerRemote. You would also need to internally delegate the actual data portal proxy to either the local proxy or the http proxy, but I'm not sure where you would do that. This is probably the most flexible option, but also the most work to implement.

**Reply by rockfordlhotka** [reply]
@kobi-shanti it doesn't sound like you need to _dynamically_ change as the app is running? But rather that you want the option to run local or remote as the app starts up?

To dynamically change, you need to determine the dynamic conditions under which you'd change - like the Internet going offline - and then switch the current data portal proxy to another one. There are different ways to do this, but it doesn't sound like you need this level of complexity.

To switch for a given launch of the application, it used to be that you'd probably change the config file. These days the choice of configuration is normally done within `Program.cs` where you set up your DI services.

CSLA 5 was in the awkward spot where we partially supported/used DI and hadn't yet fully committed to DI. Which is to say that the most common way to configure CSLA 5 was to use the fluent configuration API.

```c#
      CslaConfiguration.Configure()
        .DataPortal().DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy), "https://localhost:44332/api/dataportal");

      Csla.ApplicationContext.User = new Csla.Security.UnauthenticatedPrincipal();
```

As long as that `DataPortal` configuration occurs before any use of the data portal, that line of code can be invoked at any time.

In other words, you could pop up a dialog as the app starts, asking the user if they want to run in 2-tier or 3-tier mode, and set the proxy accordingly.

I _think_ you might be able to use that code to change the default proxy later too - I just don't remember. I know we did enable that at some point.

**Reply by kobi-shanti** [reply]
Ideally, we were hoping to dynamically change the data portal proxy after startup, when the app is running. This would mean we as developers could quickly compare performance of 2-tier vs 3-tier without needing to change our config files and restart the application. It would also allow some of our users to run in 2-tier, and others in 3-tier, based on a saved user setting.

Based on what you've both said and some tests, it doesn't look like it is possible to implement both, at least without some effort (e.g. separate RunLocal methods or custom data portal classes). From what I can tell when using the fluent configuration API as above, once a call is made to the data portal, it appears to be locked into whatever proxy was used first. We can probably get around this for the user setting component by making a non-CSLA call directly to the database to get the preferred mode (2-tier vs 3-tier), then setting the proxy in our Program.cs based on the result. 

Thanks @hurcane and @rockfordlhotka for the responses.

**Reply by rockfordlhotka** [reply]
See if there's a static method on the DataPortal type to update the current proxy.

**Reply by rockfordlhotka** [reply]
@kobi-shanti 

I finally had some time to look through the CSLA 5 codebase.

The secret in CSLA 5 is to change the `ApplicationContext.DataPortalProxy` value:

```c#
  ApplicationContext.DataPortalProxy = typeof(your proxy type).AssemblyQualifiedName;
```

or 

```c#
  ApplicationContext.DataPortalProxy = "Local";
```

Here's a link to the [code in ApplicationContext](https://github.com/MarimerLLC/csla/blob/e4855c628fc1aa3ea1d7af0afd1302b425e6b5f5/Source/Csla.Shared/ApplicationContext.cs#L459).

**Reply by kobi-shanti** [reply]
@rockfordlhotka thanks for coming back and especially for the follow up. I had a brief look following your initial response regarding a static method with no luck. We'll have a go with the ApplicationContext value approach.
